yt_dlp/extractor/crunchyroll.py

```diff
@@ -35,74 +44,177 @@ def _perform_login(self, username, password):
+        hardsub_preference = qualities(requested_hardsubs[::-1])
+        formats = []
+        for stream_type, format_id, hardsub_lang, stream_url in available_formats.values():
+            if stream_type.endswith('hls'):
+                if hardsub_lang.lower() in full_format_langs:
+                    adaptive_formats = self._extract_m3u8_formats(
+                        stream_url, display_id, 'mp4', m3u8_id=format_id,
+                        fatal=False, note=f'Downloading {format_id} HLS manifest')
+                else:
+                    adaptive_formats = (self._m3u8_meta_format(stream_url, ext='mp4', m3u8_id=format_id),)
+            elif stream_type.endswith('dash'):
+                adaptive_formats = self._extract_mpd_formats(
+                    stream_url, display_id, mpd_id=format_id,
+                    fatal=False, note=f'Downloading {format_id} MPD manifest')
             else:
-                grant_type, key = 'client_id', 'anonClientId'
-
-            initial_state, app_config = self._get_embedded_json(self._download_webpage(
-                f'https://www.crunchyroll.com/{lang}', None, note='Retrieving main page'), None)
-            api_domain = app_config['cxApiParams']['apiDomain'].replace('beta.crunchyroll.com', 'www.crunchyroll.com')
-
-            auth_response = self._download_json(
-                f'{api_domain}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
-                headers={
-                    'Authorization': 'Basic ' + str(base64.b64encode(('%s:' % app_config['cxApiParams'][key]).encode('ascii')), 'ascii')
-                }, data=f'grant_type={grant_type}'.encode('ascii'))
-            policy_response = self._download_json(
-                f'{api_domain}/index/v2', None, note='Retrieving signed policy',
-                headers={
-                    'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']
-                })
-            cms = policy_response.get('cms_web')
-            bucket = cms['bucket']
-            params = {
-                'Policy': cms['policy'],
-                'Signature': cms['signature'],
-                'Key-Pair-Id': cms['key_pair_id']
+                self.report_warning(f'Encountered unknown stream_type: {stream_type!r}', display_id, only_once=True)
+                continue
+            for f in adaptive_formats:
+                if f.get('acodec') != 'none':
+                    f['language'] = audio_locale
+                f['quality'] = hardsub_preference(hardsub_lang.lower())
+            formats.extend(adaptive_formats)
+
+        return formats
+
+    def _extract_subtitles(self, data):
+        return {
+            subtitle.get('locale'): [{
+                'url': subtitle.get('url'),
+                'ext': subtitle.get('format'),
+            }] for subtitle in traverse_obj(data, ((None, 'meta'), 'subtitles', ..., {dict}))
+        }
```

Comment:
What if multiple items have same locale? Use the setdefault pattern.
```suggestion
        subtitles = {}
        for sub in traverse_obj(data, ((None, 'meta'), 'subtitles', ..., {dict})):
            subtitles.setdefault(sub.get('locale') or 'und', []).append(
                traverse_obj(sub, {'url', 'url', 'ext': 'format'})
        return subtitles
```

Reply:
They cannot. `traverse_obj(data, ((None, 'meta'), 'subtitles'` is a `dict`. Maybe we should add a filter in there? 
```suggestion
        return {
            subtitle.get('locale'): [{
                'url': subtitle.get('url'),
                'ext': subtitle.get('format'),
            }] for subtitle in traverse_obj(data, ((None, 'meta'), 'subtitles', {dict}, ..., {dict}))
        }
```

Reply:
I don't understand. `locale` is just a key inside the final dict. Any number of dicts can have the same key, irrespective of where it is being extracted from. Anyhow, if u r sure, you can keep the code - just change to `[traverse_obj(sub, {'url', 'url', 'ext': 'format'})]`

Reply:
The format of the data is as follows:
```json
{
  "subtitles": {
    "de-DE": {
      "locale": "de-DE",
      "url": "...",
      "format": "ass"
    },
    ...
  }
}
```
Since they use locale ias both the key and provide it as data, I dont think there can be any overlap without changing the structure anyways

Reply:
I made the code be explicit about it instead. Since JSON keys can only be strings by definition, this method can never cause problems